p2p/transport/webrtc/transport.go

```diff
@@ -0,0 +1,465 @@
+	addr, wrtcComponent := ma.SplitLast(addr)
+	isWebrtc := wrtcComponent.Equal(ma.StringCast("/webrtc"))
+	if !isWebrtc {
+		return nil, fmt.Errorf("must listen on webrtc multiaddr")
+	}
+	nw, host, err := manet.DialArgs(addr)
+	if err != nil {
+		return nil, fmt.Errorf("listener could not fetch dialargs: %w", err)
+	}
+	udpAddr, err := net.ResolveUDPAddr(nw, host)
+	if err != nil {
+		return nil, fmt.Errorf("listener could not resolve udp address: %w", err)
+	}
+
+	socket, err := net.ListenUDP(nw, udpAddr)
+	if err != nil {
+		return nil, fmt.Errorf("could not listen on udp: %w", err)
+	}
+
+	// construct multiaddr
+	listenerMultiaddr, err := manet.FromNetAddr(socket.LocalAddr())
+	if err != nil {
+		_ = socket.Close()
+		return nil, err
+	}
+
+	listenerFingerprint, err := t.getCertificateFingerprint()
+	if err != nil {
+		_ = socket.Close()
+		return nil, err
+	}
+
+	encodedLocalFingerprint, err := encodeDTLSFingerprint(listenerFingerprint)
+	if err != nil {
+		_ = socket.Close()
+		return nil, err
+	}
+
+	certMultiaddress, err := ma.NewMultiaddr(fmt.Sprintf("/webrtc/certhash/%s", encodedLocalFingerprint))
+	if err != nil {
+		_ = socket.Close()
+		return nil, err
+	}
+
+	listenerMultiaddr = listenerMultiaddr.Encapsulate(certMultiaddress)
+
+	return newListener(
+		t,
+		listenerMultiaddr,
+		socket,
+		t.webrtcConfig,
+	)
+}
+
+func (t *WebRTCTransport) Dial(ctx context.Context, remoteMultiaddr ma.Multiaddr, p peer.ID) (tpt.CapableConn, error) {
+	scope, err := t.rcmgr.OpenConnection(network.DirOutbound, false, remoteMultiaddr)
+	if err != nil {
+		return nil, err
+	}
+	err = scope.SetPeer(p)
+	if err != nil {
+		return nil, err
+	}
+	pc, conn, err := t.dial(ctx, scope, remoteMultiaddr, p)
+	if err != nil {
+		scope.Done()
+		if pc != nil {
+			_ = pc.Close()
+		}
+		return nil, err
+	}
+	return conn, nil
+}
+
+func (t *WebRTCTransport) dial(
+	ctx context.Context,
+	scope network.ConnManagementScope,
+	remoteMultiaddr ma.Multiaddr,
+	p peer.ID,
+) (*webrtc.PeerConnection, tpt.CapableConn, error) {
+	var pc *webrtc.PeerConnection
+
+	remoteMultihash, err := decodeRemoteFingerprint(remoteMultiaddr)
+	if err != nil {
+		return pc, nil, fmt.Errorf("could not decode fingerprint: %w", err)
+	}
+	remoteHashFunction, ok := getSupportedSDPHash(remoteMultihash.Code)
+	if !ok {
+		return pc, nil, fmt.Errorf("unsupported hash function: %w", nil)
+	}
+
+	rnw, rhost, err := manet.DialArgs(remoteMultiaddr)
+	if err != nil {
+		return pc, nil, fmt.Errorf("could not generate dial args: %w", err)
+	}
+
+	raddr, err := net.ResolveUDPAddr(rnw, rhost)
+	if err != nil {
+		return pc, nil, fmt.Errorf("could not resolve udp address: %w", err)
+	}
+
+	// Instead of encoding the local fingerprint we
+	// instead generate a random uuid as the connection ufrag.
+	// The only requirement here is that the ufrag and password
+	// must be equal, which will allow the server to determine
+	// the password using the STUN message.
+	ufrag := "libp2p+webrtc+v1/" + genUfrag(32)
```

Comment:
TODO: given we always use this prefix, we might as well pre-allocate this as part of the 32 byte string,
and already use it in the front

Reply:
fixed in https://github.com/libp2p/go-libp2p/pull/1999/commits/f76dfc499857bffe287bedd931eeba6c652a4825 (first commit of new PR https://github.com/libp2p/go-libp2p/pull/1999)